### PR TITLE
Add derivative information to thermal expansion function objects

### LINKED ISSUE #18308: Add derivative information to AD thermal expansion materials
## Reason
Often mean thermal expansion coefficients are only dependent on temperature. These coefficients are defined and passed around as Functions with temperature passed in to the `t` (time) parameter. The method `timeDerivative` in functions can optionally be defined to give the derivative of the function, and provide derivative information to the AD methods.

## Design
Add `timeDerivative` to compute thermal expansion coefficient derivatives.

## Impact
Better usability.

### PR DESCRIPTION
**Pull request by tophmatthews**
closes #18308

### CODE REVIEW

File: modules/tensor_mechanics/src/materials/ADComputeInstantaneousThermalExpansionFunctionEigenstrain.C [hunk 1]
```diff
@@ -55,17 +55,18 @@ ADComputeInstantaneousThermalExpansionFunctionEigenstrain::computeThermalStrain(
...
 
   const Point p;
-  const Real alpha_current_temp = _thermal_expansion_function.value(current_temp, p);
+  ADReal alpha_current_temp = _thermal_expansion_function.value(current_temp, p);
+  alpha_current_temp.derivatives() =
+      _thermal_expansion_function.timeDerivative(current_temp, p) * temp.derivatives();
   const Real alpha_old_temp = _thermal_expansion_function.value(old_temp, p);
 
-  thermal_strain = old_thermal_strain + delta_T * 0.5 * (alpha_current_temp + alpha_old_temp);
+  thermal_strain = _thermal_strain_old[_qp] + delta_T * 0.5 * (alpha_current_temp + alpha_old_temp);
```
Comment (tophmatthews) [inline modules/tensor_mechanics/src/materials/ADComputeInstantaneousThermalExpansionFunctionEigenstrain.C:70]: I still get poor jacobian evalulations here @dschwen, did I do this correct? Mean thermal expansion is working great.

File: modules/tensor_mechanics/src/materials/ADComputeInstantaneousThermalExpansionFunctionEigenstrain.C [hunk 2]
```diff
@@ -55,17 +55,18 @@ ADComputeInstantaneousThermalExpansionFunctionEigenstrain::computeThermalStrain(
     _step_one = false;
 
   const Real & current_temp = MetaPhysicL::raw_value(_temperature[_qp]);
-
-  const Real & old_thermal_strain = _thermal_strain_old[_qp];
+  const ADReal & temp = _temperature[_qp];
 
   const Real & old_temp =
       (_step_one ? MetaPhysicL::raw_value(_stress_free_temperature[_qp]) : _temperature_old[_qp]);
   const Real delta_T = current_temp - old_temp;
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ADComputeInstantaneousThermalExpansionFunctionEigenstrain.C:62]: shouldn't the delta be `ADReal`?

File: modules/tensor_mechanics/src/materials/ADComputeInstantaneousThermalExpansionFunctionEigenstrain.C [hunk 3]
```diff
@@ -55,17 +55,18 @@ ADComputeInstantaneousThermalExpansionFunctionEigenstrain::computeThermalStrain(
     _step_one = false;
 
   const Real & current_temp = MetaPhysicL::raw_value(_temperature[_qp]);
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ADComputeInstantaneousThermalExpansionFunctionEigenstrain.C:57]: yeah, remove raw_value here. That's dropping derivatives

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ADComputeInstantaneousThermalExpansionFunctionEigenstrain.C:57]: No, this isn't used in derivatives, temp is. (Side note, I couldn't figure out how to get derivatives from `_temperature[_qp]` ...`_temperature[_qp].derivatives()`?)

File: modules/tensor_mechanics/src/materials/ADComputeInstantaneousThermalExpansionFunctionEigenstrain.C (hunk 2 above)
Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ADComputeInstantaneousThermalExpansionFunctionEigenstrain.C:62]: Oh ya, this is probably it

File: modules/tensor_mechanics/src/materials/ADComputeInstantaneousThermalExpansionFunctionEigenstrain.C (hunk 3 above)
Comment (dschwen) [reply modules/tensor_mechanics/src/materials/ADComputeInstantaneousThermalExpansionFunctionEigenstrain.C]: Just to be clear, the two things (raw_value and delta_T) are intimately connected...

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ADComputeInstantaneousThermalExpansionFunctionEigenstrain.C]: Yeah I think I fixed it now, take a look at the updated code.

**Review by dschwen — APPROVED** [review]
Looks good to me!

**Pull request merged**